0000-ilp-ping-protocol/0000-ilp-ping-protocol.md

```diff
@@ -0,0 +1,78 @@
+ 
+The ping protocol is not designed to debug routing issues and does not provide additional diagnostic information about the state of routing. That use case would be better served by a separate `traceroute` mechanism.
+
+All Interledger implementations _SHOULD_ respond to ping protocol requests, unless a node has a specific reason not to, such as security or privacy concerns.
+
+## Overview
+This RFC defines two ping protocol modes: unidirectional and bidirectional.
+
+* **Unidirectional**: A ping Initiator sends a specially formed ILP Prepare packet, and the Receiver either fulfills or rejects the packet according to the rules defined in [Unidirectional Mode](#Unidirectional Mode).
+* **Bidirectional**: An Initiator sends a specially formed "ping" packet that prompts the Receiver to send an additional "pong" packet. Upon receiving this pong packet, the Initiator either fulfills or rejects according to the rules defined in [Bidirectional Mode](#Bidirectional Mode), thus prompting the Receiver to fulfill or reject the original ping packet.
+
+The primary difference between the two modes is that unidirectional mode allows an Initiator to test connectivity from the Initiator to the Recipient only, whereas bidirectional mode allows the Initiator to test connectivity in both directions.
+
+## Terminology
+
+* The **Initiator** is the Interledger node initiating the ping request.
+* The **Recipient** is the Interledger node responding to the Initiator's request.
+* **Source amount** is the amount debited from the sender of an ILP payment.
+* **Destination amount** is the amount credited to the Recipient of an ILP payment.
+* The **Known Preimage** is the [ASCII](https://tools.ietf.org/html/rfc20) string `pingpingpingpingpingpingpingping`, which in binary is `70696E6770696E6770696E6770696E6770696E6770696E6770696E6770696E67`. This value is used in Unidirectional mode only. 
+* The **Known Condition** is the SHA256 hash of the Known Preimage, which when encoded using Base64 is `jAC8DGFPZPfh4AtZpXuvXFe2oRmpDVSvSJg2oT+bx34=`. This value is used in Unidirectional mode only.
+
+## Unidirectional Mode
+
+1. Initiator sends a ILP Prepare packet using the following details:
+   - **Destination**: Recipient's Interledger address.
+   - **Amount**: Any amount chosen by the Initiator.
+   - **Expiry**: Any appropriate expiry.
+   - **Condition**: The Known Condition, as defined in [Terminology](#terminology).
+   
+2. Upon receiving the Prepare packet, the Recipient MUST fulfill the payment using the following information:
+   - **Fulfillment**: The Known Fulfillment, as defined above.
+
+Note that a Recipient _MAY_ reject the payment if appropriate, for example due to an insufficient amount, invalid expiry, or for other factors.
```

Comment:
Should they use any specific errors? How should the initiator interpret the different possible errors?

Reply:
I'm leaning towards not adding anything specific about error-codes here, for two reasons:

First, [IL-RFC-27](https://github.com/interledger/rfcs/blob/master/0027-interledger-protocol-4/0027-interledger-protocol-4.md#error-codes) does a pretty good job defining ILP error codes and potential meanings. In general, I've found using those pretty intuitive on their own.

Second, the happy-path here is that a ping should be fulfilled, and if it's not being fulfilled, then the two broad categories for "why" are either: standard-ILP error messages apply (e.g., F04 - the initiator didn't send enough in the packet's amount for the packet to make it)....or....the receiver has disabled "ping" requests.

That last case (Connector has disabled ping protocol requests) is one where it _might_ make sense to be proscriptive, but here again, it feels like a typical connector would just return an `F00` or an `F02`, which sort of stands on its own without extra verbiage here.

Thoughts?